src/os/bluestore/KernelDevice.cc

```diff
@@ -440,6 +487,55 @@ void KernelDevice::_aio_thread()
   dout(10) << __func__ << " end" << dendl;
 }
 
+void KernelDevice::_discard_thread()
+{
+  std::unique_lock<std::mutex> l(discard_lock);
+  std::unique_lock<std::mutex> l2(discard_finishing_lock, std::defer_lock);
+  assert(!discard_started);
+  discard_started = true;
+  discard_cond.notify_all();
+  while (true) {
+    assert(discard_finishing.empty());
+    if (discard_submitted.empty()) {
+      if (discard_stop)
+	break;
+      dout(20) << __func__ << " sleep" << dendl;
+      discard_cond.notify_all(); // for the thread trying to drain...
+      discard_cond.wait(l);
+      dout(20) << __func__ << " wake" << dendl;
+    } else {
+      discard_finishing.swap(discard_submitted);
+      l.unlock();
+      dout(20) << __func__ << " finishing" << dendl;
+      for (auto p = discard_finishing.begin();p != discard_finishing.end(); ++p) {
+	discard(p.get_start(), p.get_len());
+      }
+
+      discard_callback(discard_callback_priv, static_cast<void*>(&discard_finishing));
+      l2.lock();
+      discard_finishing.clear();
+      l2.unlock();
+      l.lock();
```

Comment:
Can I access discard_finishing interval_set without mutex?
(l.unlock() -> accessing discard_finishing -> l.lock())
At first, i didn't use l2.lock, because discard_finishing.clear() is between l.unlock() and l.lock(), which can act like memory fence.
Nevertheless, I inserted l2 lock to protect discard_finishing interval set for safety.

I didn't use the way to put discard_finishing.clear() after l.lock(), because discard_finishing.clear() can block discard queuing thread.

Thanks

Reply:
Why not use a single lock, and include a bool discard_running; next to discard_submitted.  Then you can 

while (!discard_submitted.empty() || discard_running) {
   wait
}

under that single lock for drain?

Reply:
I understand the suggestion.
With atomic variable "discard_running",
I2 lock can be removed.

I'll update the patch!

Reply:
Not sure it even needs to be atomic.. teh drain method has to take the lock anyway to check discard_submitted, so it can be a regular bool?

Reply:
Aha, the running variable is accessed within lock scope! Then, I think it doesn't have to be atomic variable as you said.

I'll rewrite that part as soon as possible. Isn't there any other suggestion for this PR?

Reply:
Nope!

I think as a follow-up we'll want to do periodic bulk discard, but that's independent of this.

Reply:
Okay. I'll implement that part with stupid allocator first, using coarse-grained lock as soon as possible. With BitmapAllocator, fine-grained locking mechanism will be introduced.

I think the background(periodic) thread will be implemented like sync_thread in FileStore.
Thanks.